src/leo_nlm_proto4_server.erl

```diff
@@ -0,0 +1,260 @@
+%%======================================================================
+%%
+%% Network Lock Manager written in Erlang
+%%
+%% Copyright (c) 2012-2015 Rakuten, Inc.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%%======================================================================
+-module(leo_nlm_proto4_server).
+-author('Wilson Li').
+
+-include("leo_nlm_lock.hrl").
+-include("leo_nlm_proto4.hrl").
+-include_lib("leo_logger/include/leo_logger.hrl").
+-include_lib("eunit/include/eunit.hrl").
+
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2]).
+
+-export([nlmproc4_null_4/2,
+         nlmproc4_test_4/3,
+         nlmproc4_lock_4/3,
+         nlmproc4_cancel_4/3,
+         nlmproc4_unlock_4/3,
+         nlmproc4_granted_4/3,
+         nlmproc4_test_msg_4/3,
+         nlmproc4_lock_msg_4/3,
+         nlmproc4_cancel_msg_4/3,
+         nlmproc4_unlock_msg_4/3,
+         nlmproc4_granted_msg_4/3,
+         nlmproc4_test_res_4/3,
+         nlmproc4_lock_res_4/3,
+         nlmproc4_cancel_res_4/3,
+         nlmproc4_unlock_res_4/3,
+         nlmproc4_granted_res_4/3,
+         nlmproc4_share_4/3,
+         nlmproc4_unshare_4/3,
+         nlmproc4_nm_lock_4/3,
+         nlmproc4_free_all_4/3
+        ]).
+
+-define(NLM4_GRANTED,               'NLM4_GRANTED').
+-define(NLM4_DENIED,                'NLM4_DENIED').
+-define(NLM4_DENIED_NOLOCKS,        'NLM4_DENIED_NOLOCKS').
+-define(NLM4_BLOCKED,               'NLM4_BLOCKED').
+-define(NLM4_DENIED_GRACE_PERIOD,   'NLM4_DENIED_GRACE_PERIOD').
+-define(NLM4_DEADLCK,               'NLM4_DEADLCK').
+-define(NLM4_ROFS,                  'NLM4_ROFS').
+-define(NLM4_STALE_FH,              'NLM4_STALE_FH').
+-define(NLM4_FBIG,                  'NLM4_FBIG').
+-define(NLM4_FAILED,                'NLM4_FAILED').
+
+%% ---------------------------------------------------------------------
+%% API
+%% ---------------------------------------------------------------------
+%% @doc Called only once from a parent rpc server process to initialize this module
+%%      during starting a leo_storage server.
+-spec(init(any()) -> {ok, any()}).
+init(_Args) ->
+    {ok, _} = leo_nlm_lock_handler:start_link([]),
```

Comment:
It'd be good to allow users to choose the specific implementation based on `leo_storage.conf`.

I think that 
Even after finishing to implement the NLM phase 2,
There is a use case using the ets version with one leo_gateway with giving up availability to improve performance
rather than the phase 2 impl.